### PR TITLE
ENT-8617: Added PostgreSQL tunables for Federated Reporting

### PR DESCRIPTION
**Pull request by nickanderson**
This change instruments tunables for `postgresql.conf` when Federated Reporting is
enabled.

The existing defaults for `shared_buffers`, `max_locks_per_transaction`, and
`log_lock_waits` have been exposed as `cfengine_enterprise_federation.postgres_config.shared_buffers`,
`cfengine_enterprise_federation.postgres_config.max_locks_per_transaction`, and
`cfengine_enterprise_federation.postgres_config.log_lock_waits` respectively.

Additionally, since the superhub common needs `checkpoint_timeout` and
`max_wal_size` adjusted, they have been added to the default
configuration with their default values and exposed as
`cfengine_enterprise_federation.postgres_config.checkpoint_timeout` and
`cfengine_enterprise_federation.postgres_config.max_wal_size` respectively.

### CODE REVIEW

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7933/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7933/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7933/

File: MPF.md [hunk 1]
```diff
@@ -1510,6 +1510,100 @@ When custom certificates are in use distributed cleanup needs to know where to f
 
 * Added in CFEngine 3.20.0, 3.18.2
 
+#### PostgreSQL Configuration
```
Comment (craigcomstock) [inline MPF.md:1513]: change section heading to "Federated Reporting PostgreSQL Server Configuration" or similar... aka mention Federated Reporting.

File: MPF.md [hunk 2]
```diff
@@ -1510,6 +1510,100 @@ When custom certificates are in use distributed cleanup needs to know where to f
...
+
+It's not uncommon to need to configure some PostgreSQL settings differently for Federated Reporting. The settings that are exposed as tunables which can be set via augments are listed here. These do not comprise all settings that may need adjusted, only those that are most commonly adjusted.
+
+##### shared_buffers
```
Comment (craigcomstock) [inline MPF.md:1517]: For all of these we might include a link to the respective postgresql doc pages so folks know things like what units can be used and such. e.g. 2560MB or 2GB etc.

File: MPF.md [hunk 3]
```diff
@@ -1510,6 +1510,100 @@ When custom certificates are in use distributed cleanup needs to know where to f
...
+
+**History:**
+
+* Added in 3.20.0, 3.18.2
+
+##### max_wal_size
+
+Sets the WAL size that triggers a checkpoint.
+
+Maximum size to let the WAL grow during automatic checkpoints. This is a soft limit; WAL size can exceed `max_wal_size` under special circumstances, such as heavy load, a failing `archive_command`, or a high `wal_keep_size` setting. If this value is specified without units, it is taken as megabytes. The default is 1 GB (`1024MB`). Increasing this parameter can increase the amount of time needed for crash recovery.
```
Comment (craigcomstock) [inline MPF.md:1575]: might want to add that for large data load scenarios a large value is recommended, such as in Federated Reporting: 20GB is a recommended starting point for anything but demo scenarios.

File: MPF.md [hunk 4]
```diff
@@ -1510,6 +1510,100 @@ When custom certificates are in use distributed cleanup needs to know where to f
...
+
+**History:**
+
+* Added in 3.20.0, 3.18.2
+
+##### checkpoint_timeout
+
+Sets the maximum time between automatic WAL checkpoints.
+
+Maximum time between automatic WAL checkpoints. If this value is specified without units, it is taken as seconds. The valid range is between 30 seconds and one day. The default is five minutes (`5min`). Increasing this parameter can increase the amount of time needed for crash recovery.
```
Comment (craigcomstock) [inline MPF.md:1593]: Also mention why you would increase this value to our FR default of 30min: because if the timeout is too frequent the wal actually grows bigger due to full page writes needed after each checkpoint for new changes. See https://www.cybertec-postgresql.com/en/checkpoint-distance-and-amount-of-wal/

TL;DR: 

> The point now is: If checkpoints are close together there are many “first times” and full pages have to be written to the WAL quite frequently. However, if checkpoints are far apart the number of full page writes will drop dramatically leading to a lot less WAL.

File: cfe_internal/enterprise/federation/federation.cf [hunk 5]
```diff
@@ -530,18 +530,40 @@ bundle agent federation_manage_files
...
+
+      "c[max_wal_size]" -> { "ENT-8617" }
+        string => ifelse( isvariable( "cfengine_enterprise_federation.postgres_config.max_wal_size"),
+                          $(cfengine_enterprise_federation.postgres_config.max_wal_size),
+                          "1GB");
+
+      "c[checkpoint_timeout]" -> { "ENT-8617" }
+        string => ifelse( isvariable( "cfengine_enterprise_federation.postgres_config.checkpoint_timeout"),
+                          $(cfengine_enterprise_federation.postgres_config.checkpoint_timeout),
+                          "5min");
```
Comment (craigcomstock) [inline cfe_internal/enterprise/federation/federation.cf:565]: I really don't know what a good value for this is, given the un-predictable nature of when feeders dump and what cycle the checkpoint timeout might be in and whether wal gets exhausted :( Maybe something higher like 30min or 40min with the idea that we want to avoid automatic checkpoints during EVERY import cycle? I'm not sure. I suppose ideally a checkpoint might occur immediately AFTER an import actually, but postgres docs, https://www.postgresql.org/docs/current/sql-checkpoint.html, mention not to use that in "normal operations".

File: MPF.md (hunk 3 above)
Comment (nickanderson) [reply MPF.md:1575]: Since these tunables are only relevant in the context of federated reporting superhub I figured there was no need to mention that. None of these tunables will be applied if the hub is not enabled for federated reporting.

I figure if we want to suggest better defaults, we should instead change our defaults in policy. The hint that I left here is that the augments example values are using the values that we have recently suggested setting.

File: MPF.md (hunk 4 above)
Comment (nickanderson) [reply MPF.md:1593]: Since these tunables are only relevant in the context of federated reporting superhub I figured there was no need to mention that. None of these tunables will be applied if the hub is not enabled for federated reporting.

I figure if we want to suggest better defaults, we should instead change our defaults in policy. The hint that I left here is that the augments example values are using the values that we have recently suggested setting.

File: cfe_internal/enterprise/federation/federation.cf (hunk 5 above)
Comment (nickanderson) [reply cfe_internal/enterprise/federation/federation.cf:565]: Since these tunables are only relevant in the context of federated reporting superhub I figured there was no need to mention that. None of these tunables will be applied if the hub is not enabled for federated reporting.

I figure if we want to suggest better defaults, we should instead change our defaults in policy. The hint that I left here is that the augments example values are using the values that we have recently suggested setting.

File: MPF.md (hunk 1 above)
Comment (nickanderson) [reply MPF.md:1513]: Really?

I guess that both of the other headings here also explicitly say Federated Reporting, but it feels redundant given the heading levels.

Comment (vpodzime) [reply MPF.md:1513]: Yeah, it's nested properly.

File: MPF.md (hunk 3 above)
Comment (vpodzime) [reply MPF.md:1575]: Yeah, we should have good defaults instead of telling people to change our defaults to better default values.

File: cfe_internal/enterprise/federation/federation.cf [hunk 6]
```diff
@@ -530,18 +530,40 @@ bundle agent federation_manage_files
 
 bundle agent postgres_config
 # @brief Customize postgres config for superhub
+# @variable cfengine_enterprise_federation.postgres_config.shared_buffers - Sets the maximum number of locks per transaction.
```
Comment (vpodzime) [inline cfe_internal/enterprise/federation/federation.cf:533]: Shouldn't `cfengine_enterprise_federation` be a namespace? What is it otherwise?

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: cfe_internal/enterprise/federation/federation.cf (hunk 6 above)
Comment (nickanderson) [reply cfe_internal/enterprise/federation/federation.cf:533]: Good eye! Will fix.

**Comment by nickanderson** [comment]
@craigcomstock, @vpodzime please re-review

File: cfe_internal/enterprise/federation/federation.cf [hunk 7]
```diff
@@ -530,18 +530,40 @@ bundle agent federation_manage_files
...
+# @variable cfengine_enterprise_federation:postgres_config.log_lock_waits - Logs long lock waits.
+# @variable cfengine_enterprise_federation:postgres_config.max_wal_size - Sets the WAL size that triggers a checkpoint.
+# @variable cfengine_enterprise_federation:postgres_config.checkpoint_timeout- Sets the maximum time between automatic WAL checkpoints.
 {
   vars:
     am_superhub::
-      "c[shared_buffers]"
-        string => "1GB",
+      "c[shared_buffers]" -> { "ENT-8617" }
+        string => ifelse( isvariable( "cfengine_enterprise_federation.postgres_config.shared_buffers"),
```
Comment (vpodzime) [inline cfe_internal/enterprise/federation/federation.cf:542]: These should also use a namespace, AFAICT.

Comment (nickanderson) [reply cfe_internal/enterprise/federation/federation.cf]: bah, should have project wide search replace. DONE now.

**Review by vpodzime — COMMENTED** [review]
Looks good to me, but squash the commits, please.

**Comment by nickanderson** [comment]
Squashed.

**Review by vpodzime — APPROVED** [review]
ACK
